src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.cs

```diff
@@ -17,101 +17,150 @@ public static partial class JsonSerializer
                     else if (tokenType == JsonTokenType.EndObject)
                     {
-                        if (state.Current.Drain)
+                        if (readStack.Current.Drain)
                         {
-                            state.Pop();
+                            readStack.Pop();
                         }
-                        else if (state.Current.IsProcessingDictionary)
+                        else if (readStack.Current.IsProcessingDictionary)
                         {
-                            HandleEndDictionary(options, ref reader, ref state);
+                            HandleEndDictionary(options, ref reader, ref readStack);
                         }
                         else
                         {
-                            HandleEndObject(options, ref reader, ref state);
+                            HandleEndObject(options, ref reader, ref readStack);
                         }
                     }
                     else if (tokenType == JsonTokenType.StartArray)
                     {
-                        if (!state.Current.IsProcessingValue)
+                        if (!readStack.Current.IsProcessingValue)
                         {
-                            HandleStartArray(options, ref reader, ref state);
+                            HandleStartArray(options, ref reader, ref readStack);
                         }
-                        else if (HandleValue(tokenType, options, ref reader, ref state))
+                        else if (!HandleObjectAsValue(tokenType, options, ref reader, ref readStack, ref initialState))
                         {
-                            continue;
+                            // Need more data
+                            break;
                         }
                     }
                     else if (tokenType == JsonTokenType.EndArray)
                     {
-                        if (HandleEndArray(options, ref reader, ref state))
-                        {
-                            continue;
-                        }
+                        HandleEndArray(options, ref reader, ref readStack);
                     }
                     else if (tokenType == JsonTokenType.Null)
                     {
-                        if (HandleNull(ref reader, ref state, options))
-                        {
-                            continue;
-                        }
+                        HandleNull(ref reader, ref readStack, options);
                     }
                 }
             }
             catch (JsonReaderException e)
             {
                 // Re-throw with Path information.
-                ThrowHelper.ReThrowWithPath(e, state.JsonPath);
+                ThrowHelper.ReThrowWithPath(e, readStack.JsonPath);
             }
 
+            readStack.BytesConsumed += reader.BytesConsumed;
             return;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
I would imagine this method being too large to inline. We should remove the attribute.

Reply:
Really? It isn't _that_ large. Want to keep the logic together and not have to pass 6 arguments for speed. I can pull the ReadAhead case into another call and inline the outer one if it's really a problem.